### ISSUE TITLE
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.

### ISSUE THREAD

**Issue by albaldi** [jfurrow/flood#377]
Hi, I'm using rtorrent on archlinux [1]. 
Added this two lines to my rtorrent.rc as explained[2]
```
encoding_list = UTF-8
scgi_port = localhost:5000
```

and started with` screen -S torrent rtorrent`

```
netstat -tulanp | grep rtorrent
[...]
tcp        0      0 0.0.0.0:56890           0.0.0.0:*               LISTEN      31241/rtorrent      
tcp        0      0 127.0.0.1:5000          0.0.0.0:*               LISTEN      31241/rtorrent  
[...]
```
cloned your repo and builded a docker image
```
$ git clone --depth 1 https://github.com/jfurrow/flood.git
$ sudo docker build -t rtorrent-flood .

```
then I run rtorrent-flood like this:
```
$ sudo docker run -t --rm --name flood -e FLOOD_SECRET=DONTKNOWWHYCHANGEIT -e RTORRENT_SOCK=false -e FLOOD_ENABLE_SSL=false -e RTORRENT_SCGI_HOST=127.0.0.1  -e RTORRENT_SCGI_PORT=5000 -p 3000:3000 rtorrent-flood 

npm info it worked if it ends with ok
npm info using npm@4.2.0
npm info using node@v7.8.0
npm info lifecycle flood@1.0.0~prestart: flood@1.0.0
npm info lifecycle flood@1.0.0~start: flood@1.0.0

> flood@1.0.0 start /usr/src/app
> node server/bin/www
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
GET / 200 1261.967 ms - 662
GET /style.css 200 40.269 ms - -
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
GET /favicon-16x16.png 200 3.579 ms - 886
GET /favicon-32x32.png 200 2.550 ms - 1462
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
GET / 304 378.519 ms - -
GET /style.css 304 15.302 ms - -
GET /app.js 200 13.879 ms - -
GET /auth/verify?1496910193394 200 11.099 ms - 20
GET /fonts/Roboto-regular/Roboto-regular.woff2 200 5.074 ms - 10292
GET /fonts/Roboto-500/Roboto-500.woff2 200 5.312 ms - 10248
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
GET /favicon-16x16.png 304 3.597 ms - -
GET /favicon-32x32.png 304 3.305 ms - -
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
POST /auth/register 200 672.906 ms - 22
GET /fonts/Roboto-italic/Roboto-italic.woff2 200 22.738 ms - 11500
GET /fonts/Roboto-700/Roboto-700.woff2 200 5.054 ms - 10276
GET /api/history?snapshot=fiveMin 200 61.304 ms - 43
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
GET /api/stats 500 60.871 ms - 100
GET /api/client/torrents 200 53.543 ms - 2
GET /api/notifications?limit=10&start=0 200 56.476 ms - 60
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
GET /api/client/settings 500 110.768 ms - 100
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
GET /api/settings 200 97.609 ms - 2
GET /favicon-16x16.png 304 12.958 ms - -
GET /api/client/torrents/taxonomy 200 3.461 ms - 37
GET /api/history?snapshot=fiveMin 304 4.000 ms - -
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
GET /api/stats 500 16.245 ms - 100
GET /api/notifications?limit=10&start=0 304 5.691 ms - -
GET /api/client/torrents 304 2.241 ms - -
GET /api/client/torrents/taxonomy 304 2.226 ms - -

```

any way to help me debug the problem?

thanks

[1] https://git.archlinux.org/svntogit/community.git/tree/trunk/PKGBUILD?h=packages/rtorrent
[2]  https://github.com/rakshasa/rtorrent/wiki/RPC-Setup-XMLRPC

**Comment by midzer** [comment]
I experience same issue when rTorrent is auto-started as a systemd daemon, too.

Although `systemctl status rtorrent` says rTorrent is running, there are connection problems. I have to reboot system (several times) until somehow it is working magically.

Note:
rTorrent has been installed via https://github.com/Kerwood/Rtorrent-Auto-Install using ruTorrent before which had same connection problems as flood.

**Comment by albaldi** [comment]
To be sure is not a rtorrent problem I tried:
`sudo pacman -S lighttpd`

edit _/etc/lighttpd/lighttpd.conf_ and added 
```
server.modules += ( "mod_scgi" )
scgi.server = ( "/RPC2" =>
    ( "127.0.0.1" =>
        (
            "host" => "127.0.0.1",
            "port" => 5000,
            "check-local" => "disable",
            "disable-time" => 0,
        )
    )
)
```

after that
```
$ sudo systemctl start lighttpd.service 
$ sudo systemctl status lighttpd.service 
● lighttpd.service - Lighttpd Web Server
   Loaded: loaded (/usr/lib/systemd/system/lighttpd.service; disabled; vendor preset: disabled)
   Active: active (running) since Thu 2017-06-08 11:33:08 CEST; 3min 34s ago
 Main PID: 3413 (lighttpd-angel)
    Tasks: 2 (limit: 4915)
   Memory: 1.2M
      CPU: 17ms
   CGroup: /system.slice/lighttpd.service
           ├─3413 /usr/bin/lighttpd-angel -D -f /etc/lighttpd/lighttpd.conf
           └─3414 /usr/bin/lighttpd -D -f /etc/lighttpd/lighttpd.conf

Jun 08 11:33:08 bishop systemd[1]: Started Lighttpd Web Server.

```

and it works:
```
$ xmlrpc localhost get_upload_rate 
Result:

64-bit integer: 76800

```

Tried again
```
$ sudo docker run -t --rm --name flood -e FLOOD_SECRET=cosadovreimettere -e RTORRENT_SOCK=false -e FLOOD_ENABLE_SSL=false -e RTORRENT_SCGI_HOST=127.0.0.1 -e RTORRENT_SCGI_PORT=5000 -p 3000:3000 rtorrent-flood 
npm info it worked if it ends with ok
npm info using npm@4.2.0
npm info using node@v7.8.0
npm info lifecycle flood@1.0.0~prestart: flood@1.0.0
npm info lifecycle flood@1.0.0~start: flood@1.0.0

> flood@1.0.0 start /usr/src/app
> node server/bin/www

Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.
Connection refused at 127.0.0.1:5000. Check these values in config.js and ensure that rTorrent is running.

```

So I am thinking is something network related.
Now to access fllod on port 3000 on docker image it needed to be mapped to a local port. Maybe to let docker access local port 5000 I need to do something similar?

sorry, first time with docker


By the way what is **_FLOOD_SECRET_**?

**Comment by mikkoc** [comment]
This helped me to solve the same issue I was having: https://stackoverflow.com/questions/24319662/from-inside-of-a-docker-container-how-do-i-connect-to-the-localhost-of-the-mach

Basically, you need to configure rTorrent to bind not to localhost but the docker0 interface IP:
``ifconfig docker0``
And then run the container with that IP: ``-e RTORRENT_SCGI_HOST=172.x.y.z``

**Comment by jfurrow** [comment]
Hey all, unfortunately I don't have enough information to help you all debug your individual setups.

@albaldi: Also unfortunately, I didn't create Flood's `Dockerfile` and I'm not terribly familiar with Docker, so I don't now what's wrong with your setup. Can you run Flood locally without Docker and see if it can access your local instance of rTorrent? If so, then I would suspect the flood docker container can't access the host network, and I'm not sure how to fix this (maybe you just set the network mode to host? ¯\_(ツ)_/¯). I suspect the `lighttpd` proxy is part of the problem, and I'm sure you can configure it to work with a `lighttpd` proxy, but I don't know how to do it.

@midzer Since neither Flood nor ruTorrent can connect, I feel confident in deducing that rTorrent is misconfigured to accept external commands. Have you configured either `scgi_local` (for sockets) or `scgi_port` in `rtorrent.rc`? If so, what are these values, and what have you defined in Flood's `config.js`?

The error `Connection refused at 127.0.0.1:5000. Check these values in config.js` means Flood is running as expected (so far) but can't connect to rTorrent. You have to make sure Flood can access rTorrent at the IP and port specified if you're going to use the SCGI interface.

@mikkoc Points out a good thing to check, please follow his advice and check the `rtorrent.rc` configuration to see which host (if any) is being bound.

**Comment by jfurrow** [comment]
@albaldi `FLOOD_SECRET` refers to the value in `config.js` that is used for generating the JWT: https://github.com/themikenicholson/passport-jwt#configure-strategy

**Comment by albaldi** [comment]
thanks @jfurrow @mikkoc your suggestion was right.
```
$ ifconfig docker0
docker0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 172.17.0.1  netmask 255.255.0.0  broadcast 0.0.0.0
        inet6 fe80::42:93ff:fe07:b0cf  prefixlen 64  scopeid 0x20<link>
        ether 02:42:93:07:b0:cf  txqueuelen 0  (Ethernet)
        RX packets 871872  bytes 90596646 (86.3 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 1299540  bytes 1776197427 (1.6 GiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```
rtorrent must bind to docker bridge ip, because obviously 127.0.0.1/localhost refer to the container itself 
```
grep scgi ~/.rtorrent.rc 
scgi_port = 172.17.0.1:5000

```

If I wanted to use localhost `scgi_port = localhost:5000` I could run docker with `--network host` instead of default `--network bridge`

Using  `scgi_port = 0.0.0.0:5000` or  `scgi_port = :5000` is not [recommend](https://github.com/rakshasa/rtorrent/wiki/RPC-Setup-XMLRPC) .

> This is however not recommend as rtorrent has no access control, which means the http server is responsible for handling that. Anyone who can send rtorrent xmlrpc commands is likely to have the ability to execute code with the privileges of the user running rtorrent.

Thanks again. Now I only need to fix mediainfo not working #401
